pkg/monitor/health/cheker.go

```diff
@@ -0,0 +1,154 @@
+// are purely for type constraints.
+func New[T unikornv1.ManagableResourceInterface, L Lister[T]](client client.Client, driver cd.Driver, t T, l L) *Checker[T, L] {
+	return &Checker[T, L]{
+		client: client,
+		driver: driver,
+		t:      t,
+		l:      l,
+	}
+}
+
+// resourceIdentifierFromResource takes our manageable resource type and returns
+// a CD resource ID to identify its applications.
+func resourceIdentifierFromResource(r unikornv1.ManagableResourceInterface) (*cd.ResourceIdentifier, error) {
+	labels, err := r.ResourceLabels()
+	if err != nil {
+		return nil, err
+	}
+
+	id := &cd.ResourceIdentifier{
+		Labels: make([]cd.ResourceIdentifierLabel, 0, len(labels)),
+	}
+
+	for k, v := range labels {
+		id.Labels = append(id.Labels, cd.ResourceIdentifierLabel{
+			Name:  k,
+			Value: v,
+		})
+	}
+
+	return id, nil
+}
+
+// convertHealthStatus translates from the CD interface to the Kubernetes API.
+func convertHealthStatus(status cd.HealthStatus) (corev1.ConditionStatus, unikornv1.ConditionReason, string) {
+	switch status {
+	case cd.HealthStatusUnknown:
+		return corev1.ConditionUnknown, unikornv1.ConditionReasonUnknown, "unable to poll application status"
+	case cd.HealthStatusHealthy:
+		return corev1.ConditionTrue, unikornv1.ConditionReasonHealthy, "resource applications healthy"
+	case cd.HealthStatusDegraded:
+		return corev1.ConditionFalse, unikornv1.ConditionReasonDegraded, "one or more resource applications are degraded"
+	}
+
+	// NOTE: the linter will warn about non-exhaustive switches.
+	return corev1.ConditionUnknown, unikornv1.ConditionReasonUnknown, "unreachable code reached"
+}
+
+// check does the actual check for a resource and updates its status.
+func (c *Checker[T, L]) check(ctx context.Context, r unikornv1.ManagableResourceInterface) error {
+	// Grab the overall health status.
+	id, err := resourceIdentifierFromResource(r)
+	if err != nil {
+		return err
+	}
+
+	// TODO: we only support argo now, but will need an abstraction down the line.
+	// There is precedent in the main controllers.
+	healthStatus, err := c.driver.GetHealthStatus(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	updated, ok := r.DeepCopyObject().(unikornv1.ManagableResourceInterface)
+	if !ok {
+		return fmt.Errorf("%w: unable to deep copy manageable resource", ErrTypeConversion)
+	}
+
+	// And finally set the status condition.
+	status, reason, message := convertHealthStatus(healthStatus)
+
+	updated.StatusConditionWrite(unikornv1.ConditionHealthy, status, reason, message)
+
+	if err := c.client.Status().Patch(ctx, updated, client.MergeFrom(r)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Check does the actual check as described for the health checker type.
+func (c *Checker[T, L]) Check(ctx context.Context) error {
+	// NOTE: This looks expensive, but it's all cached by controller-runtime.
+	if err := c.client.List(ctx, c.l, &client.ListOptions{}); err != nil {
+		return err
+	}
+
+	for o := range c.l.All() {
```

Comment:
Ohhh so neat and tidy, `iter.Seq` for the win!

Reply:
Yes, it saves all those implicit indices!  Unless you want them and `iter.Seq2` is available.